sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -89,10 +89,12 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
   }
 
   /**
-   * Run some code involving `client` in a [[synchronized]] block and wrap certain
-   * exceptions thrown in the process in [[AnalysisException]].
+   * Run some code involving `client` and wrap certain exceptions thrown in the process in
+   * [[AnalysisException]]. Thread-safety is guaranteed here because methods in the `client`
+   * ([[org.apache.spark.sql.hive.client.HiveClientImpl]]) are already synchronized through
+   * `clientLoader` in the `retryLocked` method.
    */
-  private def withClient[T](body: => T): T = synchronized {
+  private def withClient[T](body: => T): T = {
```

Comment:
If you check the callers of `withClient`, you can find many callers conduct multiple client-related operations in the same `body`. Removing this lock might cause some concurrency issues.

Reply:
Then can we remove the synchronization of `clientLoader` in `HiveClientImpl`? If we synchronize `withClient`, then there's no need to sync each operation in `body`.

Reply:
sounds good

Reply:
but please make sure we only use the hive client here.

Reply:
Please check whether all the Hive client calls are through the Hive External Catalog. For example, `addJar` is an exception.

Reply:
I went through all methods in `HiveClient` having synchronization (except `addJar`):
- `getState`  is used only in test.
- `setOut`, `setInfo` and `setError` are only used in `SparkSQLEnv.init()`.
- all other methods are called through `HiveExternalCatalog`.

So it seems `addJar` is the only exception.

To make `addJar` also go throught `HiveExternalCatalog`, we can pass `externalCatalog` instead of `client` at [line46 in HiveSessionStateBuilder](https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionStateBuilder.scala#L46). But I don't know why we need to call `newSession()` at [line45](https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionStateBuilder.scala#L45), where a new `HiveClientImpl` instance is created, with the same class loader and Hive client.

Reply:
Last year, we had a discussion about whether `addJar` should be moved to HiveExternalCatalog. See https://github.com/apache/spark/pull/14883